File: Data Structures & Algorithms/reorder-linked-list/submission-5.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if head.next == None:
            return
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        curr = slow.next
        slow.next = None
        reversed_list = None
        while curr:
            temp = curr.next
            curr.next = reversed_list
            reversed_list = curr
            curr = temp
        begin = head
        res = begin
        begin = begin.next
        res.next = reversed_list
        reversed_list = reversed_list.next
        curr = res.next
        while begin and reversed_list:
            print(begin.val, reversed_list.val)
            if begin:
                curr.next = begin
                curr = curr.next
                begin = begin.next
            if reversed_list:
                curr.next = reversed_list
                curr = curr.next
                reversed_list = reversed_list.next
        if reversed_list:
            curr.next = reversed_list
            curr = curr.next
        if begin:
            curr.next = begin
            curr = curr.next
        curr.next = None
        return None

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-5.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length = 0;
        ListNode* curr = head;
        while (curr) {
            curr = curr->next;
            ++length;
        }
        
        int idx = length - n;
        ListNode* sentinel = new ListNode(-1);
        sentinel->next = head;
        ListNode* prev = sentinel;
        curr = head;
        while (idx) {
            prev = curr;
            curr = curr->next;
            --idx;
        }
        
        prev->next = curr->next;
        delete curr;


        return sentinel->next;
    }
};
